analysis/pseudobulk-bulk-prediction/notebooks/perform-gsea.Rmd

```diff
@@ -0,0 +1,279 @@
+---
+title: "Perform GSEA"
+author: "Stephanie J. Spielman"
+date: "`r Sys.Date()`"
+output:
+  html_notebook:
+    toc: true
+    toc_depth: 3
+    code_folding: hide
+params:
+  msigdbr_category: "H"
+  model_expr_threshold: 0.25
+  reps: 50
+---
+
+The purpose of this notebook is to run GSEA on the median of linear model residuals.
+Genes that are enriched in higher or lower residual values may point to classes genes that are preferentially detected by bulk vs single-cell RNASeq technologies.
+
+In addition, we will also run the same GSEA on shuffled gene sets (with `r params$reps`) to assess whether any results from GSEA are likely to be meaningful or noise. 
+
+
+* This notebook will use the `r params$msigdbr_category` gene sets from `MSigDB` for the GSEA.
+* This notebook will perform analyses on models from data filtered with this expression threshold: `r params$model_expr_threshold`.
+
+## Setup
+
+
+```{r setup}
+renv::load()
+
+library(ggplot2)
+library(clusterProfiler)
+theme_set(theme_bw())
+set.seed(2025)
+```
+
+### Paths
+
+```{r paths}
+model_dir <- here::here("analysis", "pseudobulk-bulk-prediction", "results", "models")
+ensembl_symbol_df <- readr::read_tsv(
+  here::here("analysis", "bulk-deconvolution", "data", "reference", "ensembl_symbol.tsv"), 
+  show_col_types = FALSE
+)
+```
+
+### Functions
+
+
+This chunk defines a function to create a data frame with a shuffled version of genesets.
+```{r}
+generate_shuffle <- function(geneset_count_df, ensembl_ids){
```

Comment:
I think my preferred approach would be to have this function take the `geneset_df` and number of replicates and return _all of the random genesets_ instead of getting passed to `purrr::map()`. That strikes me as more straightforward to follow (and more self-contained, which are related concepts IMO) if I had to return to this after some months.